common/libs/transaction.js

```diff
@@ -29,3 +39,95 @@ export type BroadcastTransaction = {|
   rawTx: string,
   signedTx: string
 |};
+
+export async function generateTransaction(
+  node: BaseNode,
+  tx: Transaction,
+  wallet: BaseWallet,
+  token: ?Token
+): Promise<BroadcastTransaction> {
+  // Reject bad addresses
+  if (!isValidETHAddress(tx.to)) {
+    return Promise.reject(new Error(translate('ERROR_5')));
+  }
+
+  // Reject token transactions without data
+  if (token && !tx.data) {
+    return Promise.reject(new Error('Tokens must be sent with data'));
+  }
+
+  // Reject gas limit under 21000 (Minimum for transaction)
+  // Reject if limit over 5000000
+  // TODO: Make this dynamic, the limit shifts
+  const limitBig = new Big(tx.gasLimit);
+  if (limitBig.lessThan(21000)) {
+    return Promise.reject(
+      new Error(translate('Gas limit must be at least 21000 for transactions'))
+    );
+  }
+
+  if (limitBig.greaterThan(5000000)) {
+    return Promise.reject(new Error(translate('GETH_GasLimit')));
+  }
+
+  // Reject gas over 1000gwei (1000000000000)
+  const gasPriceBig = new Big(tx.gasPrice);
+  if (gasPriceBig.greaterThan(new Big('1000000000000'))) {
+    return Promise.reject(
+      new Error(
+        'Gas price too high. Please contact support if this was not a mistake.'
+      )
+    );
+  }
+
+  // Ensure their balance exceeds the amount they're sending
+  // TODO: Include gas price too, tokens should probably check ETH too
```

Comment:
Yeah, definitely. I wasn't sure if hand-rolling this was the way to go though, or if `ethereumjs-tx` might be able to lend a hand. I was also going to look up a reference implementation later to see exactly what and how I should be checking.

I also would like to learn more about how much the actual broadcast portion of this catches, and how nice those errors look. I would be more inclined to lighten the load of this function if more will be caught at send time. That'll also make offline transactions less of a pain to handle, if we do less talking to the node. v3's implementation seems quite light wrt pre-send error checking.